### PR TITLE
Make CreateConsumerGroup work with StartFrom with any unit

### PR DESCRIPTION
**Pull request by KiranRG**
Since 'startFrom' does not specify a unit, make sure it works with any of the various units.

### CODE REVIEW

File: services/frontendhost/frontend.go [hunk 1]
```diff
@@ -303,13 +303,38 @@ func convertCGZoneConfigToInternal(cgZoneCfg *c.ConsumerGroupZoneConfig) *shared
 
 // convertCreateCGRequestToInternal converts Cherami CreateConsumerGroupRequest to internal shared CreateConsumerGroupRequest
 func convertCreateCGRequestToInternal(createRequest *c.CreateConsumerGroupRequest) *shared.CreateConsumerGroupRequest {
+
+	// detect and correct the units for 'startFrom' (expected internally to be in nanoseconds)
+	startFrom := createRequest.GetStartFrom()
```
Comment (venkat1109) [inline services/frontendhost/frontend.go]: can you move this logic into a helper (that takes a int64 and returns nanos)... That way, you can unit test that function separately (preferably table-test)

File: services/frontendhost/frontend_test.go [hunk 2]
```diff
@@ -663,6 +663,72 @@ func (s *FrontendHostSuite) TestFrontendHostCreateConsumerGroup() {
...
+// TestFrontendHostCreateConsumerGroupStartFrom tests whether CreateConsumerGroup is able to
+// detect and appropriately convert various startFrom units.
+func (s *FrontendHostSuite) TestFrontendHostCreateConsumerGroupStartFrom() {
+
+	testPath := uuid.New()
+	testCG := s.generateKey("/bar/CGName")
+	frontendHost, ctx := s.utilGetContextAndFrontend()
+
+	// set start-from to, say, four weeks ago
+	startFromTime := time.Now().Add(-4 * 7 * 24 * time.Hour)
```
Comment (venkat1109) [inline services/frontendhost/frontend_test.go]: need more test cases around boundary..

File: services/frontendhost/frontend_test.go [hunk 3]
```diff
@@ -663,6 +663,72 @@ func (s *FrontendHostSuite) TestFrontendHostCreateConsumerGroup() {
...
+		frontendHost.writeCacheDestinationPathForUUID(destinationUUID(cgDesc.GetDestinationUUID()), testPath)
+
+		dlqPath, _ := common.GetDLQPathNameFromCGName(testCG)
+		destDesc := &shared.DestinationDescription{
+			Path:            common.StringPtr(dlqPath),
+			DestinationUUID: common.StringPtr(uuid.New()),
+		}
+
+		// create destination is needed because of dlq destination been created at time of consumer group creation
+		s.mockController.On("CreateDestination", mock.Anything, mock.Anything).Return(destDesc, nil).Run(func(args mock.Arguments) {
```
Comment (venkat1109) [inline services/frontendhost/frontend_test.go]: see comment above, don't need any of these for testing your helper

File: services/frontendhost/frontend.go [hunk 4]
```diff
@@ -303,13 +303,38 @@ func convertCGZoneConfigToInternal(cgZoneCfg *c.ConsumerGroupZoneConfig) *shared
...
+
+	// Note: 2000-01-01T00:00 is 946684800 seconds since epoch
+	switch {
+	case startFrom == 0:
+		// special-case '0' and let it through
+
+	case startFrom < 946684800000:
```
Comment (thuningxu) [inline services/frontendhost/frontend.go]: Why not use common.MaxUint32 for seconds boundary? This goes to year 2038.

Comment (GuillaumeBailey) [inline services/frontendhost/frontend.go]: > Why not use common.MaxUint32 for seconds boundary? This goes to year 2038.

We have some values that are small, but valid. First is 0, which is a special value to mean 'now'. The second is that our documentation suggests that a very small time (e.g. '1') should be used to start from the oldest available messages. Since our CLI interprets its parameters as seconds, '1' second after epoch becomes '1E9' nanoseconds after epoch. We are currently at ~ 1.49E9 seconds after epoch, so for Kiran's scheme to work in the face of our CLI's deficiencies, we need to choose our cutoff points wisely.

Alternatively, we could also amend our CLI to treat small values (e.g. < 1E9) literally, rather than as seconds after epoch, which would simplify the model here.

File: services/frontendhost/frontend.go (hunk 1 above)
Comment (KiranRG) [inline services/frontendhost/frontend.go]: Will do. Though I still think it is better to test the behaviour of "CreateConsumerGroup" itself .. not just the helper. In the future if, say, someone does away with calling the helper, the test (that only tests the helper) would still pass.

File: services/frontendhost/frontend_test.go (hunk 2 above)
Comment (KiranRG) [inline services/frontendhost/frontend_test.go]: i am not sure what you mean by the boundary .. but i think we need to test whether this works okay with dates within "reasonable" boundaries. i'll add some more.

File: services/frontendhost/frontend_test.go (hunk 3 above)
Comment (KiranRG) [inline services/frontendhost/frontend_test.go]: see my response above; i still think it is better to test at the CreateConsumerGroup API level.

File: services/frontendhost/frontend.go (hunk 4 above)
Comment (KiranRG) [inline services/frontendhost/frontend.go]: I verified the behaviour with CLI. If *no* start-time is specified, it uses a StartFrom of "now". But if the user explicitly specifies "0", the CLI passed down "0".

I think we should fix our documentation to not say "1", but say "0" instead.

**Comment by KiranRG** [comment]
Based on my discussion with @GuillaumeBailey, I liked his idea of using the interpretation of start-from that yields the time that is closest to "now", and use that. This change implements that.

File: services/frontendhost/frontend_test.go [hunk 5]
```diff
@@ -663,6 +663,72 @@ func (s *FrontendHostSuite) TestFrontendHostCreateConsumerGroup() {
...
+
+		frontendHost.CreateConsumerGroup(ctx, req)
+	}
+
+	// test startFrom with varying units
+	testStartFrom(startFromNanos)
+	testStartFrom(startFromMicros)
+	testStartFrom(startFromMillis)
+	testStartFrom(startFromSeconds)
+	testStartFrom(0)
```
Comment (KiranRG) [inline services/frontendhost/frontend_test.go]: I need to understand this some more.
